### ISSUE TITLE
Complete animation support

### ISSUE THREAD

**Issue by mosra** [mosra/magnum#101]
This is one thing I wanted to do since the very beginning. I'm slowly implementing various bits here and there. Making this public so we can track the (non)progress better and to have a list of requirements in case someone is willing to help.

-   [ ] Interpolation
    -   [x] N-dimensional linear interpolation (`Math::lerp()`)
    -   [ ] N-dimensional Beziers with interpolation and subdivision (do we need more features?)
    -   [x] Quaternion linear/spherical interpolation (`Math::lerp()`/`Math::slerp()`)
    -   [ ] Dual quaternion linear/spherical interpolation ([original paper](http://dcgi.felk.cvut.cz/home/zara/papers/TCD-CS-2006-46.pdf))
    -   [ ] Screw interpolation (can't find the source now, but it's about decomposing the transformation into axis and angle and then linearly interpolating both)
    -   more obscure interpolation types (logarithmic)?
    -   no, I don't want to have rotation matrix interpolation with renormaliazation (for more information reread the above paper)
    -   other curve types? (OpenGEX has TCB, is it widely used outside of 3Ds Max?)
-   [ ] Class for storing animation clips -- list of *keyframes* (time/value pairs) and animation duration
    -   Templated on value type (interpolating 1D, 2D, 3D vectors, quaternions etc.)
    -   Ability to quickly calculate the value at given time using given interpolator, ability to extrapolate
    -   Preferably stateless
-   [ ] Animation class for storing animation state for given clip
    -   Templated on clip value type
    -   Something similar to what's in `SceneGraph` now, but without the scene graph and virtual calls
    -   Start/stop/pause, play count etc.
    -   Non-linear time interpolation (OpenGEX needs that)
-   [ ] Support in importers
    -   [ ] Transformation animation API -- currently all transformations are combined together into one 4x4 matrix, need to access them also separately if some are animated
    -   [ ] Skinning data API -- list of bones, bone weights in meshes, bone objects etc.
    -   [ ] Skin animation API
    -   [ ] Transformation animation import in OpenGEX
    -   [ ] Skinning data import in OpenGEX
    -   [ ] Skin animation import in OpenGEX
    -   others (light/camera property, texture coordinate animation, color animation) later?
    -   import from other formats? (no, I'm not touching COLLADA ever again)
-   [ ] Basic (transformation) animation example -- preferably a scene imported from a file (or not? in case of more obscure usages?)
-   [ ] Skeletal animation example -- also imported from a file

Question -- should I give this more priority than Vulkan support (#91) or not? I think this would contribute to general usability of this lib far more than support for a new API, but on the other hand, having Vulkan supported since *day one* would be a great feature highlight and may bring many new enthusiast users.

**Comment by Squareys** [comment]
Very nice, I'd really like to see this in Magnum, too!

> Question -- should I give this more priority than Vulkan support (#91) or not?

As long as Vulkan is not really released yet, obviously not, but I definitely agree with you that Vulkan support would be a *huge* attraction to possible new users.
And with these new users, there might be someone who has time and motivation to focus on implementing this TODO list. Whereas implementing Vulkan takes alot of wide-spread knowledge and insight of the engine, which makes it hard task for anyone other that yourself. (Also, I believe you would have more fun implementing Vulkan support than animation support. )

**Comment by mosra** [comment]
@Squareys: I might actually have more fun with animations, Vulkan will require a lot of boring iterations in the beginning to get into some useful shape first (and lots of boring reading/learning). But yeah, you're right, implementing Vulkan support would be very hard for anyone that doesn't have deep insight into design of this engine, animations not so much.

**Comment by Squareys** [comment]
In addition to the paper mentioned in "Dual quaternion linear/spherical interpolation" (which discusses dual quaternion linear (and iterative) blending and screw interpolation (ScLERP)), [this paper](http://www.cc.gatech.edu/~jarek/papers/Screw.pdf) also suggests a method to compute screw parameters for screw interpolation.

**Comment by wivlaro** [comment]
I would vote for Animations over Vulkan. Animations would benefit people using the library and make it more appealing. Vulkan is good, sure, but I don't think it would make such a huge difference as having more animation support.

**Comment by VinnyVicious** [comment]
One thing that would be really amazing is a small interpolation example along the lines of Overgrowth's techniques: http://www.gamasutra.com/view/news/216973/Video_An_indie_approach_to_procedural_animation.php

Blending between only two keyframes at a specific time, producing a realtime animation.

**Comment by wivlaro** [comment]
:+1:  Great presentation!

**Comment by mosra** [comment]
@VinnyVicious to be frank, I thought it's always done like this :) having two, maybe three keyframes defined for each animation "clip" with some specific interpolation curve in between and then (automatically) combining all those interpolated poses together .. or am I missing something?

**Comment by VinnyVicious** [comment]
@mosra, usually, i just do the entire animation in Max and export the FBX containing it. And then play it in the engine. This leads to issues known as players skating, etc.

**Comment by wivlaro** [comment]
Yeah. It's much more common practise to just have animators do all the work and then probably use a bit of blending in the engine. Some games go further, but you'll often find that artists want more control than the engineer has time. :dancer:

**Comment by VinnyVicious** [comment]
I wanted to get rid of the skating issues and also handle edge cases, like the jump technique. Everything is much more fluid using this Overgrowth technique. I can do more poses in Max and just ask the engine to work on the cue points based on player input. 

Since i'm not very good when it comes to 3D math, i'm looking some examples on the subject. :surfer:

**Comment by klaussilveira** [comment]
A good place to start learning more about IK is this CCD study: graphics.ucsd.edu/courses/cse169_w04/welman.pdf. This is also good: http://www.cs.cmu.edu/~./15464-s13/lectures/lecture6/IK.pdf

**Comment by mosra** [comment]
I think this could define the scope of initial "ready to use" animation support pretty well: http://the12principles.tumblr.com/

**Comment by mosra** [comment]
Forgot to mention that Beziers were added in 6b7097497fd7938dd0aab7baf222e848eee99287 (#165).

**Comment by mosra** [comment]
And I guess I could also publish my Animation braindump that I did a few months ago: https://gist.github.com/mosra/0ff8443d623cc0b09efb7a7299a0f692

**Comment by mosra** [comment]
More to add to the Disney 12 principles: https://twitter.com/jrcanest/status/788441769864073216

**Comment by mosra** [comment]
Stuff's being baked in #191.

**Comment by VinnyVicious** [comment]
Amazing work!

**Comment by mosra** [comment]
Initial animation support merged in #191 and mosra/magnum-plugins#46.

**Comment by mosra** [comment]
Spline interpolation, Cubic Hermite, Catmull-Rom, TCB and other splines are being done in #267.

**Comment by mosra** [comment]
Lots of great examples here: http://foxcodex.html.xdomain.jp/index.html

**Comment by mosra** [comment]
Easing is implemented in c98a936b413ffeb9e13974b9822118b532780b45 ([docs](https://doc.magnum.graphics/magnum/namespaceMagnum_1_1Animation_1_1Easing.html)), here's another collection of eased animations: https://twitter.com/DavitMasia/status/1023862760789225472

**Comment by pixelblender** [comment]
any plan to support ozz-animation (https://github.com/guillaumeblanc/ozz-animation) maybe via integration? also seems like magnum-examples doesn't have skeletal animation example yet?

**Comment by Squareys** [comment]
@pixelblender I just recently got skeletal animation working with Magnum, all the pieces are there, but there's no example, no, and you have to write your own Skinning shaders.

**Comment by mosra** [comment]
Some pieces for skeletal animation were recently integrated in e8b64544df49ecc92ff3750236e1975c13e23ea4, d7a4f3bacdf58ee97fe0c7ce03662e68edfb9bdf, 8c24ea19d3504fc96440147b4d66004638ef7290 and mosra/magnum-plugins@31fd4f11dc809ebb0a2154427e5d659037cbbaa6. Initial bits for shader support are in #444 and #441, I still need to finish and merge those (and then there will be an example as well).

**Comment by VinnyVicious** [comment]
That's awesome! Do you think an example with a character could be added so it's easier to implement?

**Comment by PierceLBrooks** [comment]
Is there any intention to eventually support automatic vertex group weight assignment algorithms such as "Bone Glow"?

**Comment by mosra** [comment]
Oh, that's cool -- didn't know such algorithms even were a thing. Yeah, why not, that'd be something to go into `MeshTools` I think. 

Unfortunately I don't have time to implement that myself anytime soon, so contributions welcome :)

**Comment by PierceLBrooks** [comment]
The two closest things I've found on GitHub so far for skinning work are https://github.com/meshonline/Surface-Heat-Diffuse-Skinning & https://github.com/JSandusky/Urho3DProcGeom
